framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java

```diff
@@ -804,7 +804,17 @@ protected void updateAnnotationSet(
 
     AnnotatedTypeMirror atmFromStorage =
         storage.atmFromStorageLocation(rhsATM.getUnderlyingType(), annotationsToUpdate);
-    updateAtmWithLub(rhsATM, atmFromStorage);
+    // It is possible that rhsATM is a type variable but atmFromStorage is not, which
+    // would cause a ClassCastException in updateAtmWithLub(). This can
+    // happen for example when a local variable of type ? extends T is assigned to a
```

Comment:
A local variable can't have a wildcard type.  I think you when an expression has type `? extends T`.